File: python/disjoint-set.py
class DisjointSet(object):
    def __init__(self, length):
        self.parent = [i for i in range(length)]
        self.size = [1 for i in range(length)]

    def root(self, i):
        if i != self.parent[i]:
            self.parent[i] = self.root(self.parent[i])
        return self.parent[i]

    def connected(self, p, q):
        return self.root(p) == self.root(q)

    def union(self, p, q):
        x = self.root(p)
        y = self.root(q)
        if x == y:
            return
        if self.size[x] < self.size[y]:
            self.parent[x] = y
            self.size[y] += self.size[x]
        else:
            self.parent[y] = x
            self.size[x] += self.size[y]
        return

# The depth of any node x is at most logN, because its depth is increased by
# one only when the size of its tree is smaller, therefore the new tree at least
# doubles in size. However, the tree can only double at most logN times.

ds = DisjointSet(10)
ds.union(1, 2)
ds.union(3, 4)
print ds.connected(1,4)
ds.union(2, 3)
print ds.connected(1,4)

File: cpp/disjoint-set.cpp
#include <iostream>
#include <vector>

using namespace std;

class DisjointSet {
    private:
        vector<int> parent, size;

        int root(int i) {
            if (i != parent[i])
                parent[i] = root(parent[i]);
            return parent[i];
        }

    public:
        DisjointSet(int length) {
            parent.reserve(length);
            size.reserve(length);
            for (int i = 0; i < length; i++) {
                parent.push_back(i);
                size.push_back(1);
            }
        }

        bool connected(int p, int q) {
            return root(p) == root(q);
        }

        void unite(int p, int q) {
            int x = root(p);
            int y = root(q);
            if (x == y) return;
            if (size[x] < size[y]) {
                parent[x] = y;
                size[y] += size[x];
            }
            else {
                parent[y] = x;
                size[x] += size[y];
            }
        }
};

int main() {
    DisjointSet ds(10);
    ds.unite(1, 2);
    ds.unite(3, 4);
    cout << ds.connected(1,4) << endl;
    ds.unite(2, 3);
    cout << ds.connected(1,4) << endl;

    return 0;
}
